File: JavaSolutions/Tree/Medium/InorderTraversal.java
/*

94. Binary Tree Inorder Traversal

https://leetcode.com/problems/binary-tree-inorder-traversal/

*/



class InorderTraversal {

    public List<Integer> inorderTraversal(TreeNode root) {
        // DFS
        List<Integer> list = new ArrayList<>();
        helper(root, list);
        return list;
    }
    
    private void helper(TreeNode root, List<Integer> list){
        if (root != null){
            if (root.left != null){
                helper(root.left, list);
            }
            list.add(root.val);
            if (root.right != null){
                helper(root.right, list);
            }
        }
    }

    public List<Integer> inorderTraversal2(TreeNode root) {
        // Like a BFS
        List<Integer> list = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode curr = root;
        while (curr != null || !stack.isEmpty()){
            while (curr != null){
                stack.push(curr);
                curr = curr.left;                
            }
            curr = stack.pop();
            list.add(curr.val);
            curr = curr.right;
        }
        return list;
    }
}

File: Python_Solutions/Tree/Medium/Inordertraversal.py
"""
94. Binary Tree Inorder Traversal
https://leetcode.com/problems/binary-tree-inorder-traversal/

Given a binary tree, return the inorder traversal of its nodes' values.

Example:

Input: [1,null,2,3]
   1
    \
     2
    /
   3

Output: [1,3,2]

"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        self.inorder_Traversal_Helper(root, res)
        return res
    
    def inorder_Traversal_Helper(self, root: TreeNode, res: List[int]):
        if root:
            self.inorder_Traversal_Helper(root.left, res)
            res.append(root.val)
            self.inorder_Traversal_Helper(root.right, res)

